### DISCUSSION TITLE
ListAsync() on ICriteria

### DISCUSSION THREAD

**Discussion by galvin1234** [nhibernate/nhibernate-core#2819] [Q&A]
Hi, I came across a situation where I need to do ListAsync() on ICriteria, and the Method execution goes to endless, 
I just would like to understand what the below method does, or where is the implementation of createInstanceMethod() method, i didn't find any reference of it. In My Case the value of createInstanceMethod variable is coming like as below image, 

<img width="854" alt="ListAsyncOnICriteria1" src="https://user-images.githubusercontent.com/16556005/121468000-bd89ac00-c9d7-11eb-90d1-9e84fbd010e4.png">

File Path : \nhibernate-core-5.3.8\src\NHibernate\Bytecode\Lightweight\ReflectionOptimizer.cs

	public virtual object CreateInstance()
		{
			return createInstanceMethod != null ? createInstanceMethod() : null;
		}

<img width="794" alt="ListAsyncOnICriteria" src="https://user-images.githubusercontent.com/16556005/121467172-43a4f300-c9d6-11eb-99d8-aa96cf010e30.png">

**Comment by hazzik** [comment]
It is a dynamic method implemented a few lines below:

https://github.com/nhibernate/nhibernate-core/blob/204b33bc0da2a4003984de238e470014a6173d07/src/NHibernate/Bytecode/Lightweight/ReflectionOptimizer.cs#L86-L124

It generates a following code:

```csharp
() => new <Type>()
```
